agent/engine/dockerclient/dockerclientfactory.go

```diff
@@ -121,8 +135,66 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
 		err = client.Ping()
 		if err != nil {
 			log.Infof("Failed to ping with Docker version %s: %v", version, err)
+			continue
 		}
 		clients[version] = client
 	}
 	return clients
 }
+
+func findDockerVersionsfromMinMaxVersions(endpoint string) (map[DockerVersion]dockeriface.Client, error) {
+	// get a Docker client with the default supported version
+	client, err := newVersionedClient(endpoint, string(minDockerAPIVersion))
+	if err != nil {
+		return nil, errors.Wrap(err, "min API version check: error while creating default Docker client")
+	}
+
+	clientVersion, err := client.Version()
+	if err != nil {
+		return nil, errors.Wrap(err, "min API version check: error while getting client version")
+	}
+
+	clients := make(map[DockerVersion]dockeriface.Client)
+	// check if the docker.Env obj has MinAPIVersion key
+	if clientVersion.Exists(minAPIVersionKey) {
+		minAPIVersion := clientVersion.Get(minAPIVersionKey)
+		apiVersion := clientVersion.Get(apiVersionKey)
+		for _, version := range getKnownAPIVersions() {
+			dockerClient, err := getDockerClientForVersion(endpoint, string(version), minAPIVersion, apiVersion)
+			if err != nil {
+				log.Debugf("Unable to get Docker client for version %s: %v", version, err)
+				continue
+			}
+			clients[version] = dockerClient
+		}
+		return clients, nil
+	}
+	return nil, errors.New("min API version not found in client version")
+}
+
+func getDockerClientForVersion(
+	endpoint string,
+	version string,
+	minAPIVersion string,
+	apiVersion string) (dockeriface.Client, error) {
+	lessThanMinCheck := "<"+minAPIVersion
+	moreThanMaxCheck := ">"+apiVersion
+	minVersionCheck, minErr := utils.Version(version).Matches(lessThanMinCheck)
```

Comment:
I don't think you need multiple `error` variables here. If you stick to the policy of creating/initializing a variable only when needed, you'll do something like this:
```
minVersionCheck, err := utils.Version(version).Matches(lessThanMinCheck)
if err != nil {
  ...
}

maxVersionCheck, err := utils.Version(version).Matches(moreThanMaxCheck)
if err != nil {
 ...
}
if minVersionCheck || ...
```

Reply:
Earlier was doing a check of the error messages together, didn't modify it later. changed now.